File: FFT/Lab3/Solutions/OpenACC/c/cufftACC.c
#include "../../../Common/c/fileReader.h"
#include "../../../Common/c/acchelper.h"
#include <unistd.h>
//CUDA includes
#include <cufft.h>

//Size of array in bytes
extern size_t MEM_SIZE;

//Host allocated pointers
extern Complex *h_inputFFT, *h_outputFFT;

//Timing variables
extern struct timeval start, stop;
float execTime;


void initialize() {

    /**
     * Initializing the complex number
     * with a random wave using Openacc directives
     */
    float time;
    int i, j;
    double val, commonCalc, W1, W2, DELTA_W, AVG_W;
    int h_freqArray[ITERS];

    //Loading the frequencies from a file.
    loadFrequencies("../../../Common/frequencies.txt", h_freqArray, ITERS);
    
    // pcopyin is short for present_or_copyin
    // Whatever that is not on the device will be copied.
    #pragma acc data pcopyin(h_inputFFT[:SIGNAL_SIZE], h_freqArray[:ITERS])
    {
        gettimeofday(&start, NULL);
        
        #pragma acc parallel loop
        for (i=0; i<SIGNAL_SIZE; ++i) {
            val = 0.0;
            commonCalc = TWO_PI * i / NUM_SAMPLES;
            W1 = FREQUENCY * commonCalc;
            
            #pragma acc loop seq
            for (j=0; j<ITERS; ++j) {
                
                    W2  = h_freqArray[j] * commonCalc;
                    DELTA_W = fabs(W1 - W2);
                    AVG_W = (W1 + W2) * 0.5;
                    val += 2 * cos(DELTA_W * 0.5 * i) * sin(AVG_W * i);
            }
            h_inputFFT[i].x = val / ITERS;
            h_inputFFT[i].y = 0.0;
        }

        gettimeofday(&stop, NULL);
    }

    time = (float)(stop.tv_usec - start.tv_usec) / 1000000 + (float)(stop.tv_sec - start.tv_sec);
    execTime += time;
    printf("INIT:%.2f\t", time);
}

void executeFFT() {

    float time;
    //Creating a plan for Complex to Complex operation in double precision.
    cufftHandle plan;

    gettimeofday(&start, NULL);
    
    cufftPlan1d(&plan, SIGNAL_SIZE, CUFFT_Z2Z, 1); // Z2Z is C2C with Double Precision
    /**
     * CUDA libraries require the GPU allocated data. But the call to cufft is from the host.
     * host_data construct makes the pointers of GPU allocated data visible to the host.
     * use_device() clause simply tells the compiler to use the GPU copy instead of CPU copy
     */
    #pragma acc host_data use_device(h_inputFFT, h_outputFFT)
    {
        //Perform FFT
        cufftExecZ2Z(plan, (cufftDoubleComplex *)h_inputFFT,
                        (cufftDoubleComplex *)h_outputFFT,
                        CUFFT_FORWARD);

        //Perform Inverse FFT
        cufftExecZ2Z(plan, (cufftDoubleComplex *)h_outputFFT,
                        (cufftDoubleComplex *)h_outputFFT,
                        CUFFT_INVERSE);
    }

    cufftDestroy(plan);
    
    gettimeofday(&stop, NULL);
    
    time = (float)(stop.tv_usec - start.tv_usec) / 1000000 + (float)(stop.tv_sec - start.tv_sec);
    execTime += time;
    printf("FFT:%.2f\t", time);
}

void calculateDifference() {
    
    float time;
    int i;
    //Data is already present on the GPU
    // Using the output array to store the differences
    #pragma acc data pcopyin(h_inputFFT[:SIGNAL_SIZE], h_outputFFT[:SIGNAL_SIZE])
    {
        gettimeofday(&start, NULL);
        
        #pragma acc parallel loop
        for (i=0; i<SIGNAL_SIZE; ++i) {
            //Compute the difference
            //Normalize the output of the Inverse FFT by dividing it by SIZE beforehand
            //Store the differences in the output array
            h_outputFFT[i].x = fabs(h_inputFFT[i].x - (h_outputFFT[i].x / SIGNAL_SIZE));
        }
        
        gettimeofday(&stop, NULL);
    }
    
    time = (float)(stop.tv_usec - start.tv_usec) / 1000000 + (float)(stop.tv_sec - start.tv_sec);
    execTime += time;
    printf("DIFF:%.2f\t", time);
}


int main(int argc, char **argv) {

    int i, N = 1;
    // Number of times to execute the parallel portions of the code
    // To run ./fftacc.out <N> or make ARGS="<N>"
    //By default the N = 1
    if (argc > 1) { 
        N = atoi(argv[1]);
    }
    printf("\n\n==============Execution================\n\n");
    printf("SIGNAL SIZE = %d\n\n", SIGNAL_SIZE);

    //Allocating pointers
    allocate();

    printf("Times in seconds\n\n");

    //Allocating device memory before doing any operations.
    #pragma acc data create(h_inputFFT[:SIGNAL_SIZE]) copyout(h_outputFFT[:SIGNAL_SIZE])
    {
        for (i=0; i<N; ++i) {
        
            //Initializing
            initialize();
            //Executing FFT + IFFT
            executeFFT();
            //Calculating difference between inputFFT and inverse outputFFT
            calculateDifference();
            printf("TOTAL:%.2f\n\n", execTime);
		sleep(10);
        }
    }

    //printMeanError is executed on the CPU but it requires the h_outputFFT array which stores the differences.
    printMeanError(h_outputFFT);
    
    //Deallocating pointers
    deallocate();

    printf("\n\n==============Terminated================\n\n");
    return 0;
}

File: FFT/Lab3/Solutions/OpenACC/fortran/cufftACC.f90
program MAIN
    
    use ACCHELPER
    use openacc
    use cufft
    use fileReader
    implicit none

    !For CL Inputs
    include "lib3f.h"
    integer :: i, N, argc
    character*2 :: argv
    
    argc = iargc()
    call getarg(argc, argv)

    N = 1
    if (argc == 1) then
        read (argv, *) N
    end if

    print *, "==============Execution================"
    print *
    print *, "SIGNAL SIZE = ", SIGNAL_SIZE
    print *
    print *, "Times in seconds"

    !Allocate pointers
    allocate(h_inputFFT(SIGNAL_SIZE), h_outputFFT(SIGNAL_SIZE))
    
    !Allocate the device memory before calling any of the functions
    !Copying out the output array which stores the differences that
    !are used by printMeanError funciton

    !$acc data create(h_inputFFT) copyout(h_outputFFT)
    do i=1, N

        !Initializing
        call initialize()
        !Executing FFT + IFFT
        call executeFFT()
        !Calculating difference between inputFFT and inverse outputFFT
        call calculateDifference()
        
        print *, "INIT:", initTime, "FFT:", fftTime, "DIFF:", diffTime
        print *, "TOTAL:", execTime
    end do
    !$acc end data
    print *
    !printMeanError is executed on the CPU but it requires the h_outputFFT array which stores the differences
    call printMeanError(h_outputFFT)

    !Deallocating pointers
    deallocate(h_inputFFT, h_outputFFT)
    print *
    print *, "==============Terminated================"
    print *

    contains

        subroutine initialize()

            !Initializing the complex number
            !with a random wave using Openacc directives
            
            implicit none
            integer :: i, j
            real*8 :: val, commonCalc, W1, W2, DELTA_W, AVG_W
           
            !reading a file and loading the frequencies
            call loadFrequencies("../../../Common/frequencies.txt", h_freqArray, ITERS)
            
            !pcopyin stands for present or copyin
            !Already loaded data will not be copied
            !parallelize the outer loop
            
            !$acc data pcopyin(h_inputFFT, h_freqArray)
            
            call cpu_time(start)
            
            !$acc parallel loop
            outer:do i=1, SIGNAL_SIZE
                val = 0.0
                commonCalc = TWO_PI * (i-1) / NUM_SAMPLES
                W1 = FREQUENCY * commonCalc
                
                !This loop is executed sequentially by each thread
                !$acc loop seq
                inner:do j=1, ITERS
                    W2  = h_freqArray(j) * commonCalc
                    DELTA_W = DABS(W1 - W2)
                    AVG_W = (W1 + W2) * 0.5
                    val = val + 2 * COS(DELTA_W * 0.5 * (i-1)) * SIN(AVG_W * (i-1))
                end do inner
                h_inputFFT(i) = dcmplx(val / ITERS, 0.0)
            end do outer
            !$acc end parallel
            
            call cpu_time(finish)
            
            !$acc end data

            initTime = finish - start
            execTime = execTime + initTime
        end subroutine initialize

        subroutine executeFFT()
            implicit none

            integer :: plan, err
            
            call cpu_time(start)

            !Create a cufft plan for Z2Z i.e complex to complex in Double precision
            err = cufftPlan1d(plan, SIGNAL_SIZE, CUFFT_Z2Z, 1)
            
            ! CUDA libraries require the GPU allocated data. But the call to cufft is from the host.
            ! host_data construct makes the pointers of GPU allocated data visible to the host.
            ! use_device() clause simply tells the compiler to use the GPU copy instead of CPU copy
            
            !$acc host_data use_device(h_inputFFT, h_outputFFT)

                !Perform FFT out of place transform
                err = err + cufftExecZ2Z(plan, h_inputFFT, h_outputFFT, CUFFT_FORWARD)
                !Perform Inverse FFT in plance transform
                err = err + cufftExecZ2Z(plan, h_outputFFT, h_outputFFT, CUFFT_INVERSE)
            
            !$acc end host_data
            
            err = err + cufftDestroy(plan)

            call cpu_time(finish)

            fftTime = finish - start
            execTime = execTime + fftTime
        end subroutine executeFFT

        subroutine calculateDifference()
            implicit none
            
            integer :: i
            real*8 :: diff
            
            !$acc data pcopyin(h_inputFFT, h_outputFFT)
            
            call cpu_time(start)
            
            !$acc parallel loop
            do i=1, SIGNAL_SIZE
                !normalize the inverse fft output
                diff = DABS(real(h_inputFFT(i) - (real(h_outputFFT(i))/SIGNAL_SIZE)))
                !storing the difference in outputFFT
                h_outputFFT(i) = dcmplx(diff, 0)
            end do
            !$acc end parallel
            
            call cpu_time(finish)

            !$acc end data
            
            diffTime = finish - start
            execTime = execTime + diffTime
        end subroutine calculateDifference
end program MAIN
